### PR TITLE
magento/devdocs#: Routing. Action result object.

### PR DESCRIPTION
**Pull request by atwixfirster**
## Purpose of this pull request

This pull request (PR) adds information about action result objects.

## Affected DevDocs pages

- https://devdocs.magento.com/guides/v2.1/extension-dev-guide/routing.html
- https://devdocs.magento.com/guides/v2.2/extension-dev-guide/routing.html
- https://devdocs.magento.com/guides/v2.3/extension-dev-guide/routing.html

## Links to Magento source code

- [2.1](https://github.com/magento/magento2/blob/2.1/lib/internal/Magento/Framework/Controller/ResultFactory.php#L19)
- [2.2](https://github.com/magento/magento2/blob/2.2/lib/internal/Magento/Framework/Controller/ResultFactory.php#L21)
- [2.3](https://github.com/magento/magento2/blob/2.3/lib/internal/Magento/Framework/Controller/ResultFactory.php#L22)

Thank you!

whatsnew
Added [reference documentation](https://devdocs.magento.com/guides/v2.3/extension-dev-guide/routing.html#result-object) about the `result object` returned by the `Action` class

### CODE REVIEW

**Comment by devops-devdocs** [comment]
An admin must run tests on this PR before it can be merged.

File: guides/v2.1/extension-dev-guide/routing.md [hunk 1]
```diff
@@ -160,6 +160,21 @@ $this->_forward('action', 'controller', 'Other_Module')
...
 
+
+{: .bs-callout .bs-callout-info }
+Action class should return a `result object`.
+
+## Result object
+
+Name | Description
+--- | ---
+`json` | Sets `Content-Type:application/json` in the header and returns a json encoded representation of an array with data.
```
Comment (erikmarr) [inline guides/v2.1/extension-dev-guide/routing.md]: remove ending period

File: guides/v2.1/extension-dev-guide/routing.md [hunk 2]
```diff
@@ -160,6 +160,21 @@ $this->_forward('action', 'controller', 'Other_Module')
...
+`raw` | Returns the data as it's been set. Does not set a `Content-Type` of the header.
```
Comment (erikmarr) [inline guides/v2.1/extension-dev-guide/routing.md]:
```suggestion
`raw` | Returns the data as it's been set. Does not set a `Content-Type` in the header.
```
Also, remove period from end of last sentence.

File: guides/v2.1/extension-dev-guide/routing.md [hunk 3]
```diff
@@ -160,6 +160,21 @@ $this->_forward('action', 'controller', 'Other_Module')
...
+`redirect` | Creates an external redirect, which the browser follows and requests a new url.
```
Comment (erikmarr) [inline guides/v2.1/extension-dev-guide/routing.md]: remove period from end of sentence

File: guides/v2.1/extension-dev-guide/routing.md [hunk 4]
```diff
@@ -160,6 +160,21 @@ $this->_forward('action', 'controller', 'Other_Module')
...
+`forward` | Internally calls the execute method of another action class and does not trigger a new request from the browser. The URL stays the same.
```
Comment (erikmarr) [inline guides/v2.1/extension-dev-guide/routing.md]: Remove period from end of last sentence.

File: guides/v2.1/extension-dev-guide/routing.md [hunk 5]
```diff
@@ -160,6 +160,21 @@ $this->_forward('action', 'controller', 'Other_Module')
...
+`layout` | View result. A generic layout response can be used for rendering any kind of layout. So it comprises a response body from the layout elements it has and sets it to the HTTP response
```
Comment (erikmarr) [inline guides/v2.1/extension-dev-guide/routing.md]:
```suggestion
`layout` | View result. You can use a generic layout response to render any kind of layout. The layout comprises a response body from its layout elements and sets it to the HTTP response
```

File: guides/v2.3/extension-dev-guide/routing.md [hunk 6]
```diff
@@ -165,6 +165,20 @@ $this->_forward('action', 'controller', 'Other_Module')
...
 
+{: .bs-callout .bs-callout-info }
+Action class should return a `result object`.
+
+## Result object
+
+Name | Description
+--- | ---
+`json` | Sets `Content-Type:application/json` in the header and returns a json encoded representation of an array with data.
+`raw` | Returns the data as it's been set. Does not set a `Content-Type` of the header!
```
Comment (erikmarr) [inline guides/v2.3/extension-dev-guide/routing.md]: Remove ending exclamation mark

File: guides/v2.3/extension-dev-guide/routing.md [hunk 7]
```diff
@@ -165,6 +165,20 @@ $this->_forward('action', 'controller', 'Other_Module')
 {: .bs-callout .bs-callout-tip }
 Use the [`ActionFactory`] in your router to create an instance of an `Action` class.
 
+{: .bs-callout .bs-callout-info }
+Action class should return a `result object`.
+
+## Result object
```
Comment (erikmarr) [inline guides/v2.3/extension-dev-guide/routing.md]: Please apply above edits to this table as well except where noted below.

File: guides/v2.1/extension-dev-guide/routing.md [hunk 8]
```diff
@@ -160,6 +160,21 @@ $this->_forward('action', 'controller', 'Other_Module')
...
+`page` | View result. Encapsulates page type, page configuration and imposes certain layout handles. Triggers `layout.xml` rendering into HTML.
```
Comment (erikmarr) [inline guides/v2.1/extension-dev-guide/routing.md]:
```suggestion
`page` | View result. Encapsulates page type, page configuration, and imposes certain layout handles. `page` triggers `layout.xml` to render into HTML
```

File: guides/v2.3/extension-dev-guide/routing.md (hunk 6 above)
Comment (atwixfirster) [reply guides/v2.3/extension-dev-guide/routing.md]: **done**

@erikmarr , could you please review?

Thank you!

File: guides/v2.3/extension-dev-guide/routing.md (hunk 7 above)
Comment (atwixfirster) [reply guides/v2.3/extension-dev-guide/routing.md]: **done**

File: guides/v2.1/extension-dev-guide/routing.md (hunk 8 above)
Comment (atwixfirster) [reply guides/v2.1/extension-dev-guide/routing.md]: **done**

File: guides/v2.1/extension-dev-guide/routing.md (hunk 5 above)
Comment (atwixfirster) [reply guides/v2.1/extension-dev-guide/routing.md]: **done**

File: guides/v2.1/extension-dev-guide/routing.md (hunk 4 above)
Comment (atwixfirster) [reply guides/v2.1/extension-dev-guide/routing.md]: **done**

File: guides/v2.1/extension-dev-guide/routing.md (hunk 3 above)
Comment (atwixfirster) [reply guides/v2.1/extension-dev-guide/routing.md]: **done**

File: guides/v2.1/extension-dev-guide/routing.md (hunk 2 above)
Comment (atwixfirster) [reply guides/v2.1/extension-dev-guide/routing.md]: **done**

File: guides/v2.1/extension-dev-guide/routing.md (hunk 1 above)
Comment (atwixfirster) [reply guides/v2.1/extension-dev-guide/routing.md]: **done**

**Comment by erikmarr** [comment]
running tests

**Comment by erikmarr** [comment]
Thanks @atwixfirster!

**Pull request merged**